schematic

```diff
@@ -0,0 +1,304 @@
+in the migrations directory.
+
+Migrations are just files whose names start with a number, like
+`001-adding-awesome.sql`.  They're matched against `'^\d+'` so you can
+put zeros in front to keep ordering in `ls` happy, and whatever you want
+after the migration number, such as text describing the migration.
+
+If a file ends with no extension or .sql it is treated as SQL and passed
+to your database. If the file ends with an extension in handlers, the file
+is passed to that command to do with as it wishes.
+
+schematic creates a table (named in settings.py) with one column, that
+holds one row, which describes the current version of the database.  Any
+migration file with a number greater than the current version will be
+applied to the database and the version tracker will be upgraded.  The
+migration and version bump are performed in a transaction.
+
+The version-tracking table will initially be set to 0, so the 0th
+migration could be a script that creates all your tables (for
+reference).  Migration numbers are not required to increase linearly.
+
+schematic doesn't pretend to be intelligent. Running migrations manually
+without upgrading the version tracking will throw things off.
+
+Tested on sqlite and mysql.
+
+NOTE: any superfluous output, like column headers, will cause an error.
+On mysql, this is fixed by using the `--silent` parameter.
+"""
+
+import optparse
+import os
+import re
+import six
+import sys
+import time
+from subprocess import Popen, PIPE, STDOUT
+
+
+SETTINGS = 'settings'
+VARIABLES = ['db', 'table', 'handlers']
+OPTIONAL_VARIABLES = {'handlers': {}}
+IGNORE_FILES = ['.pyc']
+
+CREATE = 'CREATE TABLE %s (version INTEGER NOT NULL);'
+COUNT = 'SELECT COUNT(version) FROM %s;'
+SELECT = 'SELECT version FROM %s;'
+INSERT = 'INSERT INTO %s (version) VALUES (%s);'
+UPDATE = 'UPDATE %s SET version = %s;'
+UPGRADE = 'BEGIN;\n%s\n%s\nCOMMIT;'
+
+
+class SchematicError(Exception):
+    """Base class for custom errors."""
+
+
+def exception(f):
+    """Decorator to turn a function into an subclass of SchematicError."""
+    class E(SchematicError):
+        def __init__(self, *args, **kwargs):
+            msg = f(self, *args, **kwargs)
+            SchematicError.__init__(self, msg)
+    E.__name__ = f.__name__
+    return E
+
+
+@exception
+def MissingSettings(self):
+    return "Couldn't import settings file"
+
+
+@exception
+def SettingsError(self, k):
+    return "Couldn't find value for '%s' in %s.py" % (k, SETTINGS)
+
+
+@exception
+def DbError(self, cmd, stdout, stderr, returncode):
+    msg = '\n'.join(["Had trouble running this: %s", "stdout: %s",
+                     "stderr: %s", "returncode: %s"])
+    return msg % (cmd, stdout, stderr, returncode)
+
+
+@exception
+def MultipleMigrations(self, num):
+    return 'Multiple migrations with number: %d' % num
+
+
+@exception
+def ExternalError(self, cmd, stdout, stderr, returncode):
+    return DbError(cmd, stdout, stderr, returncode)
```

Comment:
Can we make them real classes maybe?

Reply:
https://github.com/mozilla/addons-server/pull/10537/commits/1af201b7f4c6e171bf0c57650deda745a04076d8